set of fake controllers

rhaschke opened the pull request:
This reopens #20 against jade-devel.

Implement as set of fake controllers traversing
- smoothly interpolated path
- via points only
- last point only

rhaschke commented:
Revived initialization of robot state to named posture. 
Removed hard-coded parameters from @davetcoleman's version.

rhaschke commented:
@davetcoleman Do you have any idea, why warehouse_ros fails in Jade's Travis?

davetcoleman commented:
This repo is still using the old .travis.yml file, which has some sort of problem with using rosinstall files... I don't care to support it since I spent a ton of time making moveit_ci. Would you like me to add it to this repo? It would be no prob

rhaschke commented:
Good point. I will add moveit_ci to this PR. This way it's automatically 
checked.

rhaschke commented:
@davetcoleman Cool! Checks passed with new `moveit_ci`.

rhaschke commented:
From my point of view, this PR could be reviewed and merged.

davetcoleman commented on moveit_fake_controller_manager/README.md:
```diff
@@ -0,0 +1,29 @@
+
+initial:
+  - group: arm
+    pose:  home
```
This is great documentation - however I think it should be in `doc/tutorial.rst` and built with doxygen then linked from moveit website like the rest of moveit's documentation.

rhaschke replied:
I think I can manage to create a `tutorial.rst` additional to `README.md`. However, I'm not sure how to build that with doxygen. I found though a file conf.py in those repos. How is this picked up by doxygen during build?

davetcoleman replied:
I just went through this process for moveit_planners - in the /ompl folder you'll see:
- conf.py
- rosdoc.yaml
- doc/moveit-theme/*

You can mostly copy paste those here. Then just edit `doc/tutorial.rst` for this plugin.

v4hn replied:
This is not about doxygen.
`conf.py` and `*.rst` files are part of the Sphinx documentation,
which is also build through rosdoc.

davetcoleman replied:
Ah yes, I used the wrong word - I meant Sphinx

davetcoleman commented on moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp:
```diff
@@ -236,14 +92,22 @@ class MoveItFakeControllerManager : public moveit_controller_manager::MoveItCont
+        else if (type == "via points")
+          controllers_[name].reset(new ViaPointController(name, joints, pub_));
+        else
+          controllers_[name].reset(new InterpolatingController(name, joints, pub_));
```
i prefer to check for exact string ("interpolate") and have the else statement throw an error in case user did something odd like misspell "last point" - they wouldn't expect to get "interpolate"

davetcoleman commented on moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp:
```diff
@@ -252,6 +116,75 @@ class MoveItFakeControllerManager : public moveit_controller_manager::MoveItCont
+
+    if (param.getType() != XmlRpc::XmlRpcValue::TypeArray || param.size() == 0)
+    {
+      ROS_ERROR_ONCE_NAMED("loadInitialJointValues", "Parameter 'initial' should be an array of (group, pose) structs.");
```
i love that these are `_NAMED` +1!

davetcoleman commented on moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp:
```diff
@@ -324,15 +260,19 @@ class MoveItFakeControllerManager : public moveit_controller_manager::MoveItCont
-} // end namespace moveit_fake_controller_manager
+}  // end namespace moveit_fake_controller_manager
 
+#include <pluginlib/class_list_macros.h>
```
shouldn't this include still be at top of file?

rhaschke replied:
Does need to. As this is only required for declaration of the plugin, I prefer to have them together.
This way, you cannot forget to (re)move the include if the plugin declaration is (re)moved.

davetcoleman replied:
I think you could make that argument for all sorts of things that are included. still, elsewhere in moveit and most all other code i've ever seen its always at the top of the file

rhaschke replied:
Not for many things actually. However, I don't mind and will change this.

davetcoleman commented on moveit_fake_controller_manager/src/moveit_fake_controllers.cpp:
```diff
@@ -0,0 +1,288 @@
+
+bool LastPointController::waitForExecution(const ros::Duration &)
+{
+  ros::Duration(0.5).sleep(); // give some time to receive the published JointState
```
Is this really necessary?

rhaschke replied:
Hm. It was there before. I removed it and it failed to update to the new pose.
However, https://github.com/ros-planning/moveit_ros/pull/716 it might be superfluous now.

davetcoleman commented on moveit_fake_controller_manager/src/moveit_fake_controllers.cpp:
```diff
@@ -0,0 +1,288 @@
+
+ThreadedController::~ThreadedController()
+{
+  thread_.join();
```
should this also have `cancel_ = true`?

rhaschke replied:
You might want to wait for the trajectory to finish, wouldn't you?

davetcoleman replied:
but this would cause the program to hang if its a long trajectory, eventually roslaunch will just kill it

rhaschke replied:
Valid argument. I will cancel the trajectory then.

davetcoleman commented on moveit_fake_controller_manager/src/moveit_fake_controllers.cpp:
```diff
@@ -0,0 +1,288 @@
+  // no further interpolation
+  ros::Time startTime = ros::Time::now();
+  for (std::vector<trajectory_msgs::JointTrajectoryPoint>::const_iterator
+       via = t.joint_trajectory.points.begin(), end = t.joint_trajectory.points.end(); !cancelled() && via != end; ++via) {
```
line break for {

davetcoleman commented on moveit_fake_controller_manager/src/moveit_fake_controllers.cpp:
```diff
@@ -0,0 +1,288 @@
+    js.effort = via->effort;
+
+    ros::Duration waitTime = via->time_from_start - (ros::Time::now() - startTime);
+    if (waitTime.toSec() > 1e-3) {
```
turn `1e-3` into const var with comment to how it was chosen

davetcoleman commented on moveit_fake_controller_manager/src/moveit_fake_controllers.cpp:
```diff
@@ -0,0 +1,288 @@
+}
+
+
+InterpolatingController::InterpolatingController(const std::string &name, const std::vector<std::string> &joints, const ros::Publisher &pub)
```
line length

davetcoleman commented on moveit_fake_controller_manager/src/moveit_fake_controllers.cpp:
```diff
@@ -0,0 +1,288 @@
+{
+  double duration = (next.time_from_start - prev.time_from_start).toSec();
+  double alpha = duration > std::numeric_limits<double>::epsilon() ? (elapsed - prev.time_from_start).toSec() / duration
+                                                                   : 1.0;
```
IMHO the readability here is bad - why not just use an if statement?

davetcoleman commented on moveit_fake_controller_manager/src/moveit_fake_controllers.cpp:
```diff
@@ -0,0 +1,288 @@
+  double alpha = duration > std::numeric_limits<double>::epsilon() ? (elapsed - prev.time_from_start).toSec() / duration
+                                                                   : 1.0;
+  js.position.resize(prev.positions.size());
+  for (std::size_t i = 0, end = prev.positions.size(); i < end; ++i)
```
interesting way of doing a for loop - does the `end` var actually have better performance? seems like a really unusual way to do things

rhaschke replied:
That's the usual way I do it. Learned from [Effective C++](https://www.amazon.de/Effective-Specific-Addison-Wesley-Professional-Computing/dp/0321334876). The performance gain is not dramatic, but you don't need to evaluate `size()` all the time.

davetcoleman commented on moveit_fake_controller_manager/src/moveit_fake_controllers.h:
```diff
@@ -0,0 +1,124 @@
+ *********************************************************************/
+
+/* Author: Ioan Sucan, Robert Haschke */
+
```
need header guard

davetcoleman commented on moveit_fake_controller_manager/src/moveit_fake_controllers.h:
```diff
@@ -0,0 +1,124 @@
+};
+
+
+class ThreadedController : public BaseFakeController
```
I find it way easier to find code when each class is in its own file with same name. Also per ROS style guide:

> If the file primarily implements a class, name the file after the class. For example the class ActionServer would live in the file action_server.h.

Could you do that here?

rhaschke replied:
I understand your argument. On the contrary, I don't like bloating the folder with files.
We have five different classes in there, which are closely related and typically really simple. The disclaimer would take more lines than the actual code...

davetcoleman commented on moveit_fake_controller_manager/src/moveit_fake_controllers.h:
```diff
@@ -0,0 +1,124 @@
+  virtual moveit_controller_manager::ExecutionStatus getLastExecutionStatus();
+
+protected:
+  bool cancelled() { return cancel_; }
```
separate lines

davetcoleman commented:
Looks good after addressing my comments

rhaschke commented:
I addressed most of your comments in code. See my answers to the others ;-)

davetcoleman commented on moveit_fake_controller_manager/src/moveit_fake_controllers.h:
```diff
@@ -0,0 +1,132 @@
+#include <boost/thread/thread.hpp>
+
+#ifndef MOVEIT_FAKE_CONTROLLERS
+#define MOVEIT_FAKE_CONTROLLERS
```
I see why you named it this, but now I notice the header is not in the include folder. If you don't think anyone will use these classes, why use a header at all then?

rhaschke replied:
It's a clean separation of interface and implementation, which facilitates reading: I can immediately see the whole API in the header. I thought this is good programming practice.
Even though, the header is not public now, having it separated from the beginning will also facilitate making it public later.

rhaschke commented:
I addressed the remaining issues. Don't mind to squash on merging ;-)
- added rosdoc.yaml
- call cancelTrajectory() in destructor.
- moved header include to the top

davetcoleman commented:
+1 thanks!

v4hn commented:
So much for having different commits for different things...
I doubt the `.travis.yml` is part of "set of fake controllers".
Now it's one commit in `jade-devel`.

v4hn commented:
cherry-picked to kinetic-devel, as we just discussed we don't want to have things merged in older branches if they are not also merged in the most current one...

Who do I blame... I could have already written a proposed policy for pull-requests/merging...

v4hn commented:
@rhaschke could you backport this to indigo? It would be _awsome_ to have continuous fake controllers ready to use in indigo.
I would like to point some of my students (total MoveIt! newbies) to them for testing their projects.

rhaschke commented:
> @rhaschke could you backport this to indigo? It would be awsome to have continuous fake controllers ready to use in indigo.

Done (#23).
@v4hn: Porting this to kinetic, you copied `.travis.yml` as is, without switching to kinetic distro. I fixed that in kinetic branch directly.

v4hn commented:
On Wed, Jul 27, 2016 at 03:03:03PM -0700, Robert Haschke wrote:

> @v4hn: Porting this to kinetic, you copied `.travis.yml` as is, without switching to kinetic distro. I fixed that in kinetic branch directly.

Oh crap, thanks!

That's why I would have preferred it in a separate request :)

davetcoleman commented:
> I fixed that in kinetic branch directly.

that's a bending of our protocol! :worried:

davetcoleman commented:
Could you announce this great new feature on the mailing list?

v4hn commented:
On Thu, Jul 28, 2016 at 05:31:25PM -0700, Dave Coleman wrote:

> Could you announce this great new feature on the mailing list?

Do that on the next release.
The release includes other notable features too (e.g. the "cancel" button")
